envoy/network/connection_handler.h

```diff
@@ -242,5 +242,68 @@ class ActiveUdpListenerFactory {
 
 using ActiveUdpListenerFactoryPtr = std::unique_ptr<ActiveUdpListenerFactory>;
 
+/**
+ * Internal listener callbacks.
+ */
+class InternalListener : public virtual ConnectionHandler::ActiveListener {
+public:
+  /**
+   * Called when a new connection is accepted.
+   * @param socket supplies the socket that is moved into the callee.
+   */
+  virtual void onAccept(ConnectionSocketPtr&& socket) PURE;
+};
+
+using InternalListenerPtr = std::unique_ptr<InternalListener>;
+using InternalListenerOptRef = OptRef<InternalListener>;
+
+/**
+ * The query interface of the registered internal listener callbacks.
+ */
+class InternalListenerManager {
+public:
+  virtual ~InternalListenerManager() = default;
+
+  /**
+   * Return the internal listener binding the listener address.
+   *
+   * @param listen_address the internal address of the expected internal listener.
+   */
+  virtual InternalListenerOptRef
+  findByAddress(const Address::InstanceConstSharedPtr& listen_address) PURE;
```

Comment:
Naive question: can the provided `listen_address` point to a listener which is not internal?
What will be returned then?

Reply:
Seems like it's required to be an internal address at the moment (via ASSERT). It returns nullopt if it's not internal.
I haven't actually touched this code in the PR, just moved headers around to avoid loops.